### ISSUE TITLE
using php error log for operational logging

### ISSUE THREAD

**Issue by 123andy** [vanderbilt-redcap/flight-tracker#1]
The app contaminates php error logs with operational logging.  It should have its own dedicated optional logging mechanism.

**Comment by scottjpearson** [comment]
In version 2.26.1, I've disabled printing to the PHP error log if it's in the context of a particular project. (Those logs are searchable through the Flight Tracker application.) If it's acting on a system level, Flight Tracker still prints to the PHP error log as none of the projects have jurisdiction.

**Comment by 123andy** [comment]
I'd recommend having an optional system parameter for an error log and separate it from the php error log.  If the log is still getting polluted with non-errors from Flight Tracker then we still have an issue.

**Comment by scottjpearson** [comment]
Thant makes sense. All logging statements should now be associated with a pid as of 2.26.1, out at the end of this week. Thanks.

**Comment by 123andy** [comment]
@scottjpearson - Flight tracker is still writing to my server's php_errors file.  This really isn't acceptable.  In your code you need to replace all `error_log` function calls with an internal logging function (which can default to error_log if you like) but should allow me to select an alternate file/destination for flight_tracker status messages.

**Comment by scottjpearson** [comment]
@123andy The intended behavior is to write to the External Module log file. The only exceptions are (a) when it is pid-less environment (should not happen), (b) when it is in a plugin mode and not an ExtMod (only relevant to one project at Vanderbilt), (c) when the 'test' parameter is supplied in the GET params, or (d) when it is running on localhost.

A grep reveals that the only lines of code which reference error_log directly are in a handful of Vanderbilt-specific scenarios.

Can you provide an example of these statements? Thanks.

**Comment by 123andy** [comment]
Here are some example statements:

```
[26-Jan-2022 21:04:10 America/Los_Angeles] Adding summary_all_degrees
[26-Jan-2022 21:04:31 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:00:04 America/Los_Angeles] Flight Tracker running for pids ["21227","21230","24536"]
[27-Jan-2022 00:00:04 America/Los_Angeles] Checking for redcaptest in redcap.stanford.edu
[27-Jan-2022 00:02:01 America/Los_Angeles] Currently running updateBibliometrics for pid 21230 with status WAIT
[27-Jan-2022 00:02:01 America/Los_Angeles] Promoting updateBibliometrics for 21230 to RUN (6 items in batch queue; 1 records) at 2022-01-27 00:02:01
[27-Jan-2022 00:02:01 America/Los_Angeles] 21230: Updating metrics for Record 27 (38 rows)
[27-Jan-2022 00:02:03 America/Los_Angeles] Contacting https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&retmode=xml&id=33044293,32917541,32701731,31767277,31453362,31173112,31095681,31090887,31009390,30922747
[27-Jan-2022 00:02:04 America/Los_Angeles] Response code 200; 120035 bytes
[27-Jan-2022 00:02:25 America/Los_Angeles] Contacting https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&retmode=xml&id=30762599,30656563,30427983,30278971,30256923,30061041,29986015,24850730,31713994,31295766
[27-Jan-2022 00:02:25 America/Los_Angeles] Response code 200; 130846 bytes
[27-Jan-2022 00:02:47 America/Los_Angeles] Contacting https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&retmode=xml&id=33681473,33681473,31713994,31453362,31295766,33681473,31713994,31453362,31295766,33681473
[27-Jan-2022 00:02:47 America/Los_Angeles] Response code 200; 108075 bytes
[27-Jan-2022 00:03:32 America/Los_Angeles] Contacting https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&retmode=xml&id=31713994,31453362,31295766,33681473,31713994,31453362,31295766
[27-Jan-2022 00:03:33 America/Los_Angeles] Response code 200; 99090 bytes
[27-Jan-2022 00:05:05 America/Los_Angeles] Running SQL DELETE FROM redcap_data WHERE project_id = '21230' AND record = '27' AND field_name LIKE 'citation_%' AND (instance IN ('26','27','28','29','30','31','32','33','34','35'))
[27-Jan-2022 00:05:05 America/Los_Angeles] SQL: 308 rows affected
[27-Jan-2022 00:05:05 America/Los_Angeles] Running SQL DELETE FROM redcap_data WHERE project_id = '21230' AND record = '27' AND field_name LIKE 'citation_%' AND (instance IN ('36','37','38','39','40','41'))
[27-Jan-2022 00:05:05 America/Los_Angeles] SQL: 188 rows affected
[27-Jan-2022 00:05:06 America/Los_Angeles] Done with updateBibliometrics at 2022-01-27 00:05:06
[27-Jan-2022 00:06:01 America/Los_Angeles] Currently running updateBibliometrics for pid 21230 with status DONE
[27-Jan-2022 00:06:01 America/Los_Angeles] Promoting makeSummary for 21230 to RUN (5 items in batch queue; 30 records) at 2022-01-27 00:06:01
[27-Jan-2022 00:06:02 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:06:04 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:06:06 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:06:08 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:06:11 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:06:13 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:06:16 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:06:22 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:06:26 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:06:34 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:06:36 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:06:39 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:06:40 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:06:43 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:06:45 America/Los_Angeles] Done with makeSummary at 2022-01-27 00:06:45
[27-Jan-2022 00:07:01 America/Los_Angeles] Currently running makeSummary for pid 21230 with status DONE
[27-Jan-2022 00:07:01 America/Los_Angeles] Promoting makeSummary for 21230 to RUN (4 items in batch queue; 11 records) at 2022-01-27 00:07:01
[27-Jan-2022 00:07:04 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:07:06 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:07:15 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:07:21 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:07:27 America/Los_Angeles] Adding summary_all_degrees
[27-Jan-2022 00:07:29 America/Los_Angeles] Done with makeSummary at 2022-01-27 00:07:29
[27-Jan-2022 00:08:01 America/Los_Angeles] Currently running makeSummary for pid 21230 with status DONE
[27-Jan-2022 00:08:01 America/Los_Angeles] Promoting updateBibliometrics for 24536 to RUN (3 items in batch queue; 1 records) at 2022-01-27 00:08:01
[27-Jan-2022 00:08:01 America/Los_Angeles] 24536: Updating metrics for Record 27 (22 rows)
[27-Jan-2022 00:08:04 America/Los_Angeles] Contacting https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&retmode=xml&id=33044293,32917541,32701731,31767277,31453362,31173112,31095681,31090887,31009390,30922747
[27-Jan-2022 00:08:04 America/Los_Angeles] Response code 200; 120035 bytes
[27-Jan-2022 00:08:21 America/Los_Angeles] Contacting https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&retmode=xml&id=30762599,30656563,30427983,30278971,30256923,30061041,29986015,24850730,31713994,31295766
[27-Jan-2022 00:08:21 America/Los_Angeles] Response code 200; 130846 bytes
[27-Jan-2022 00:08:38 America/Los_Angeles] Contacting https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&retmode=xml&id=33681473
[27-Jan-2022 00:08:38 America/Los_Angeles] Response code 200; 3200 bytes
[27-Jan-2022 00:09:09 America/Los_Angeles] Done with updateBibliometrics at 2022-01-27 00:09:09
[27-Jan-2022 00:10:02 America/Los_Angeles] Currently running updateBibliometrics for pid 24536 with status DONE
[27-Jan-2022 00:10:02 America/Los_Angeles] Promoting makeSummary for 24536 to RUN (2 items in batch queue; 30 records) at 2022-01-27 00:10:02
[27-Jan-2022 00:10:02 America/Los_Angeles] Done with makeSummary at 2022-01-27 00:10:02
[27-Jan-2022 00:11:01 America/Los_Angeles] Currently running makeSummary for pid 24536 with status DONE
[27-Jan-2022 00:11:01 America/Los_Angeles] Promoting makeSummary for 24536 to RUN (1 items in batch queue; 11 records) at 2022-01-27 00:11:01
[27-Jan-2022 00:11:01 America/Los_Angeles] Done with makeSummary at 2022-01-27 00:11:01
[27-Jan-2022 00:13:01 America/Los_Angeles] Sending Flight Tracker email for pid 21230 to jpmeyer@stanford.edu, eorasa@stanford.edu
[27-Jan-2022 00:13:02 America/Los_Angeles] Sending Flight Tracker email for pid 24536 to jpmeyer@stanford.edu, eorasa@stanford.edu
```

**Comment by scottjpearson** [comment]
By my current trace, the $pid should have been present and should have prevented and error_log calls. However, to be doubly sure, I just wrapped the error_log calls in the central logging function in a

> if (self::isVanderbilt()) {
>     error_log(...);
> }

construct. This will come out in FT v4.0.3 and higher. Will make a note to check in next week with you to confirm/deny success. Thanks for reporting.

**Comment by scottjpearson** [comment]
@123andy Did this issue get resolved with last weekend's release (4.0.3)?

**Comment by 123andy** [comment]
Just installed it today -- give me a little time to see if anything new appears...
________________________________
